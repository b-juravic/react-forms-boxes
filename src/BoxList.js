import React, { useState } from 'react';
import NewBoxForm from './NewBoxForm'
import Box from './Box';
import { v4 as uuid } from 'uuid';

function BoxList() {
  // create boxes state
  const [boxes, setBoxes] = useState([]);

  // remove a box from state when its button is pressed
  const removeBox = id => {
    setBoxes(boxes.filter(box => box.id !== id));
  };

  // add a new box to state
  const addBox = box => {
    let newBox = { ...box, id: uuid() }
    setBoxes(boxes => [ ...boxes, newBox ]);
  };

  // render all boxes in state
  const renderBoxes = () => {
    return (
      <div>
        {boxes.map(box => (
          <Box
            backgroundColor={box.color}
            width={box.width}
            height={box.height}
            key={box.id}
            id={box.id}
            handleClose={removeBox}
          />
        ))}
      </div>
    );
  };

  return (
    <div>
      <NewBoxForm addBox={addBox} />
      {renderBoxes()}
    </div>
  );
}

export default BoxList;