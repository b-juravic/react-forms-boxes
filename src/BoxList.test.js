import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", function() {
  render(<BoxList />);
});

it("matches snapshot", function() {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("should render a box", function() {
  const { asFragment, queryByText, getByLabelText, getByText, queryByAttribute } = render(<BoxList />);
  
  // make sure no boxes start on screen
  expect(queryByText('X')).not.toBeInTheDocument();

  //grab inputs
  const widthInput = getByLabelText("Box Width");
  const heightInput = getByLabelText("Box Height");
  const colorInput = getByLabelText("Box Color");
  const submitButton = getByText("Submit");

  // fill out form and submit
  fireEvent.change(widthInput, { target: { value: "20px" }});
  fireEvent.change(heightInput, { target: { value: "20px" }});
  fireEvent.change(colorInput, { target: { value: "blue" }});
  fireEvent.submit(submitButton);

  // test!
  expect(queryByText('X')).toBeInTheDocument();
  expect(asFragment()).toMatchSnapshot();
});

it("should be able to remove a box", function() {
  const { asFragment, queryByText, getByLabelText, getByText, queryByAttribute } = render(<BoxList />);

  //grab inputs
  const widthInput = getByLabelText("Box Width");
  const heightInput = getByLabelText("Box Height");
  const colorInput = getByLabelText("Box Color");
  const submitButton = getByText("Submit");

  // fill out form and submit
  fireEvent.change(widthInput, { target: { value: "20px" }});
  fireEvent.change(heightInput, { target: { value: "20px" }});
  fireEvent.change(colorInput, { target: { value: "blue" }});
  fireEvent.submit(submitButton);

  // make sure box is there
  expect(queryByText('X')).toBeInTheDocument();
  
  // grab remove button and fire
  const removeButton = getByText("X");
  fireEvent.click(removeButton);
  
  // test!
  expect(queryByText('X')).not.toBeInTheDocument();

});