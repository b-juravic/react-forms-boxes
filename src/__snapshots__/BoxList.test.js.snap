// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="width"
      >
        Box Width
      </label>
      <br />
      <input
        id="width"
        name="width"
        value=""
      />
      <br />
      <label
        for="height"
      >
        Box Height
      </label>
      <br />
      <input
        id="height"
        name="height"
        value=""
      />
      <br />
      <label
        for="color"
      >
        Box Color
      </label>
      <br />
      <input
        id="color"
        name="color"
        value=""
      />
      <br />
      <button>
        Submit
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;

exports[`should render a box 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="width"
      >
        Box Width
      </label>
      <br />
      <input
        id="width"
        name="width"
        value=""
      />
      <br />
      <label
        for="height"
      >
        Box Height
      </label>
      <br />
      <input
        id="height"
        name="height"
        value=""
      />
      <br />
      <label
        for="color"
      >
        Box Color
      </label>
      <br />
      <input
        id="color"
        name="color"
        value=""
      />
      <br />
      <button>
        Submit
      </button>
    </form>
    <div>
      <div>
        <div
          style="background-color: blue; width: 20px; height: 20px; display: inline-block;"
        />
        <button>
          X
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
